File: zig/1-blinky-bm/main/memory.zig
const std = @import("std");
const idf = @import("esp_idf");
const sys = idf.sys;

const utils = @import("utils.zig");

pub fn report(tasks: []const sys.TaskHandle_t) void {
    for (tasks) |task| {
        const name = sys.pcTaskGetName(task);

        const task_stack_usage = stackUsage(task) catch |err| {
            std.log.err("Failed to get stack usage of task {s}: {}", .{ name, err });
            continue;
        };

        std.log.info("{s} stack usage: {} B", .{ name, task_stack_usage });
    }

    const total_heap_size = sys.heap_caps_get_total_size(0);
    const free_heap_size = sys.heap_caps_get_free_size(0);
    const heap_usage = total_heap_size - free_heap_size;
    std.log.info("Heap usage: {} B", .{heap_usage});
}

fn stackUsage(task: sys.TaskHandle_t) !u32 {
    var snapshot: sys.TaskSnapshot_t = undefined;
    try utils.rtosCheckError(sys.vTaskGetSnapshot(task, &snapshot));

    return snapshot.pxEndOfStack - snapshot.pxTopOfStack;
}

File: c/1-blinky-bm/main/memory.c
#include "esp_log.h"
#include "esp_private/freertos_debug.h"
#include "freertos/idf_additions.h"

#include "memory.h"

static const char *TAG = "memory";

int32_t stack_usage(TaskHandle_t task) {
  BaseType_t err;

  TaskSnapshot_t snapshot;
  err = vTaskGetSnapshot(task, &snapshot);
  if (err != pdTRUE) {
    ESP_LOGE(TAG, "vTaskGetSnapshot fail (0x%x)", err);
    return -1;
  }

  return snapshot.pxEndOfStack - snapshot.pxTopOfStack;
}

void memory_report() {
  const TaskHandle_t task = xTaskGetCurrentTaskHandle();

  char *name = pcTaskGetName(task);

  int32_t task_stack_usage = stack_usage(task);
  if (task_stack_usage == -1) {
    ESP_LOGE(TAG, "stack_usage fail");
    return;
  }

  ESP_LOGI(TAG, "%s stack usage: %" PRIi32 " B", name, task_stack_usage);

  size_t total_heap_size = heap_caps_get_total_size(0);
  size_t free_heap_size = heap_caps_get_free_size(0);
  size_t heap_usage = total_heap_size - free_heap_size;
  ESP_LOGI(TAG, "Heap usage: %d B", heap_usage);
}
